### PR TITLE
Connections API

### LINKED ISSUE #3291: Implement new "connections" model and API
## Current behavior

We have a "databases" API which has developed some cruft and does not suit our changes in direction for v0.1.4.

## Desired behavior

We should have a new "connections" API which behaves as follows

- Use the endpoint `/api/db/v0/connections/`
- Have a "nickname" field that can be changed. Nicknames must be unique across all connections within one Mathesar installation.
- Not have `deleted` or `editable` fields, as in the current "databases" API.
- Write and read connections _only_ to and from the internal (Django) database. (This is in contrast to the old databases API which considered the environment variables to be a second source of truth for user database connections.)

### PR DESCRIPTION
**Pull request by mathemancer**
Fixes #3291 

This PR is an initial iteration towards replacing the "Database" concept in our codebase with a "DB Connection" (usually just "Connection") concept.

**Technical details**

The following endpoints are changed:

- `/api/ui/v0/databases/*` -> `/api/ui/v0/connections/*`
- `/api/db/v0/databases/*` -> `/api/db/v0/connections/*`

Note that as indicated by the wildcard, all endpoints under these namespaces are affected.

Changes to keys in JSON representations of connections (orig databases):

- `name` -> `nickname`
- `db_name` -> `database`
- `deleted` -> _(removed)_
- `editable` -> _(removed)_
- `error` -> _(removed)_

Changes to common data JSON keys:

- `.current_db` -> `.current_db_connection`
- `.databases` -> `.connections`
- `.databases.[i].name -> `.connections[0].nickname`
- `.databases.[i].db_name -> `.connections[0].database`
- `.internal_database` -> `.internal_db_connection`
- `.internal_database.db_name` -> `.internal_db_connection.database`

Note that in order to keep this PR small, the Schema model refers to its `database` rather than `connection` in the API and common data.

Behavior changes:

When creating or updating a connection (formerly database) via POST or PATCH, we no longer test the actual connection or install Mathesar on the DB. That will be handled by a subsequent PR on the RPC endpoint.

Regarding `MATHESAR_DATABASES`:

I left that variable for now, since removing it would require rewriting major pieces of our back end test suite. However, I don't think it will affect much. I think we should remove that variable from our docs and examples, _except_ for the dev setup, where having a preinstalled DB automatically created is convenient.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by mathemancer** [comment]
@Anish9901 Please review, but don't merge until @seancolsen makes the required front end changes.

File: mathesar/api/serializers/databases.py
Comment (Anish9901) [inline mathesar/api/serializers/databases.py:26]: There is a minor bug here, the uri formed will always be namespaced under `ui/` as the `basename` is same for both the `db` & `ui` route. Not sure how significant it is for the frontend though.

**Review by Anish9901 — APPROVED** [review]
Looks great @mathemancer! I've addressed a small bug pleas take a look.

**Comment by pavish** [comment]
I'm taking over the frontend work from @seancolsen for changes needed in this PR.

File: mathesar/api/serializers/databases.py
Comment (mathemancer) [reply mathesar/api/serializers/databases.py:26]: Hmmm.  I'll experiment with it a bit.

**Comment by pavish** [comment]
@mathemancer @Anish9901 I've fixed the breaking changes on the frontend, except the previous 'Database Connections' settings page which will be rewritten as per the new spec. This PR can be merged when ready.

File: mathesar/api/serializers/databases.py
Comment (mathemancer) [reply mathesar/api/serializers/databases.py:26]: @Anish9901 I poked around and found that the behavior's the same as before. I agree that this is a bit of a bug, but it's apparently harmless. Given that we're going to be completely overhauling the API structure in the near future, I want to avoid fixing it for now.

**Pull request merged**
